oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/session/InMemoryOidcProviderSessionRegistry.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2002-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.oidc.authentication.session;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.security.oauth2.client.oidc.authentication.logout.LogoutTokenClaimNames;
+import org.springframework.security.oauth2.client.oidc.authentication.logout.OidcLogoutToken;
+
+/**
+ * An in-memory implementation of {@link OidcProviderSessionRegistry}
+ *
+ * @author Josh Cummings
+ * @since 6.1
+ */
+public final class InMemoryOidcProviderSessionRegistry implements OidcProviderSessionRegistry {
```

Comment:
I'm not sure that `SessionRegistry` is the right fit. If we are using it for storing the provider's session id, then the concern is calling `getAllSessions`, which returns a `List`. I really don't like the idea of having to load everything into memory, and while I recognize that for many, the list returned will be small, it pays off for us to use `Iterable` for interfaces where there is a good likelihood that there will be database-based implementations out there.

Further, `SessionRegistryImpl#refreshLastRequest` feels super weird in this case. Are we supporting the client refreshing the provider's session? The implementation seems quite wrong and we'd at least need a provider-specific implementation that throws an exception for this method's implementation. I don't like having meaningless information hanging around in-memory and worse having an API that updates and changes that meaningless information.

Lastly, I find this implementation quite nice in that I don't have to rematerialize an `OidcUser` given a backchannel request. I can just treat `OidcLogoutToken` as a first-class citizen.

Reply:
`OidcProviderSessionRegistry` would only be limited to OIDC logout. We do need to address the "polling" behaviour implemented by concurrency control and ideally we have one interface that can be reused for both use cases. As well, Spring Authorization Server is using `SessionRegistry` for tracking OIDC sessions and we should also consider what SAML logout would use too.

A single interface flexible enough for all 4x use cases is the ideal solution. I do feel `SessionRegistry` is generic enough to fit all these use cases. I think it would be a super valuable exercise to give `SessionRegistry` a try so we can work through the API issues and see if we can improve it. Can you give this a try?

> the concern is calling `getAllSessions`, which returns a `List`

Suggested improvement to address this can be:

```java
@Deprecated
List<SessionInformation> getAllSessions(Object principal, boolean includeExpiredSessions);

default Iterator<SessionInformation> getSessions(Object principal, boolean includeExpiredSessions) {
	return getAllSessions(principal, includeExpiredSessions).iterator();
}
```